src/mono/mono/metadata/threads.c

```diff
@@ -3912,18 +3915,58 @@ mono_thread_init_apartment_state (void)
 		thread->apartment_state = ThreadApartmentState_Unknown;
 	}
 #endif
+
+#if TARGET_OSX || HOST_MACCAT || TARGET_IOS || TARGET_TVOS
+
+	ERROR_DECL (error);
+	MONO_STATIC_POINTER_INIT (MonoMethod, create_autoreleasepool)
+
+		create_autoreleasepool = mono_class_get_method_from_name_checked (mono_defaults.thread_class, "CreateAutoreleasePool", 0, 0, error);
+		mono_error_assert_ok (error);
+
+	MONO_STATIC_POINTER_INIT_END (MonoMethod, create_autoreleasepool)
+
+	if (mono_runtime_get_no_exec ())
+		return;
+
+	mono_runtime_try_invoke_handle (create_autoreleasepool, NULL_HANDLE, NULL, error);
+	mono_error_cleanup (error);
+
+#endif
+
+	thread->flags |= MONO_THREAD_FLAG_PLATFORM_INIT;
 }
 
 void 
-mono_thread_cleanup_apartment_state (void)
+mono_thread_cleanup_platform_state (void)
 {
-#ifdef HOST_WIN32
 	MonoInternalThread* thread = mono_thread_internal_current ();
+	if (!(thread->flags & MONO_THREAD_FLAG_PLATFORM_INIT))
+		return;
```

Comment:
nit: `CoInitializeEx` is always called, but `CoUninitialize` is guarded by the flag.  They should be balanced.

Reply:
I don't see how that is true &ndash; at least from `mono_thread_init_platform_state`. The last thing that `mono_thread_init_platform_state` does is flip this bit so when `mono_thread_cleanup_platform_state` is called it will run or simply return if it wasn't run.

Reply:
yea, you're right.  the early return in init is only on Darwin, and the CoInitializeEx is only on Windows.